transaction_mpx.go

```diff
@@ -0,0 +1,198 @@
+package dbr
+
+import (
+	"context"
+	"database/sql"
+	"time"
+)
+
+// TxMpx is a multiplexed transaction created by SessionMpx.
+type TxMpx struct {
+	PrimaryTx   *Tx
+	SecondaryTx *Tx
+	SecondaryQ  *Queue
+}
+
+func (txMpx *TxMpx) AddJob(job *Job) error {
+	return txMpx.SecondaryQ.AddJob(job)
+}
+
+func (txMpx *TxMpx) PrimaryExecContext(ctx context.Context, query string, args ...interface{}) (sql.Result, error) {
+	return txMpx.PrimaryTx.ExecContext(ctx, query, args...)
+}
+
+func (txMpx *TxMpx) SecondaryExecContext(ctx context.Context, query string, args ...interface{}) (sql.Result, error) {
+	return txMpx.SecondaryTx.ExecContext(ctx, query, args...)
+}
+
+func (txMpx *TxMpx) PrimaryQueryContext(ctx context.Context, query string, args ...interface{}) (*sql.Rows, error) {
+	return txMpx.PrimaryTx.QueryContext(ctx, query, args...)
+}
+
+func (txMpx *TxMpx) SecondaryQueryContext(ctx context.Context, query string, args ...interface{}) (*sql.Rows, error) {
+	return txMpx.SecondaryTx.QueryContext(ctx, query, args...)
+}
+
+// BeginTxs creates a transaction with TxOptions.
+func (smpx *SessionMpx) BeginTxs(ctx context.Context, opts *sql.TxOptions) (*TxMpx, error) {
+	to := smpx.GetTimeout()
+
+	primaryTx, err := smpx.PrimaryConn.BeginTx(ctx, opts)
+	if err != nil {
+		return nil, smpx.PrimaryConn.EventErr("dbr.primary.begin.error", err)
+	}
+	smpx.PrimaryConn.Event("dbr.primary.begin")
+
+	txmPx := &TxMpx{
+		PrimaryTx: &Tx{
+			EventReceiver: smpx.PrimaryEventReceiver,
+			Dialect:       smpx.PrimaryConn.Dialect,
+			Tx:            primaryTx,
+			Timeout:       to,
+		},
+		SecondaryTx: &Tx{
+			EventReceiver: smpx.SecondaryEventReceiver,
+			Dialect:       smpx.SecondaryConn.Dialect,
+			Timeout:       to,
+		},
+		SecondaryQ: smpx.secondaryQ,
+	}
+
+	j := NewJob("dbr.secondary.begin", nil, func() error {
+		secondaryTx, rerr := smpx.SecondaryConn.BeginTx(ctx, opts)
+		if rerr != nil {
+			return rerr
+		}
+		txmPx.SecondaryTx.Tx = secondaryTx
+		return nil
+	})
+	smpx.AddJob(j)
+
+	return txmPx, nil
+}
+
+// Begin creates a multiplexed transaction for the given session.
+func (smpx *SessionMpx) Begin() (*TxMpx, error) {
+	return smpx.BeginTxs(context.Background(), nil)
+}
+
+func (txMpx *TxMpx) GetTimeout() time.Duration {
+	return txMpx.PrimaryTx.GetTimeout()
+}
+
+func (txMpx *TxMpx) Exec(query string, args ...interface{}) (sql.Result, error) {
+	return txMpx.ExecContext(context.Background(), query, args...)
+}
+
+func (txMpx *TxMpx) ExecContext(ctx context.Context, query string, args ...interface{}) (sql.Result, error) {
+	j := NewJob("dbr.secondary.exec_context", map[string]string{"sql": query}, func() error {
+		if txMpx.SecondaryTx.Tx == nil {
+			panic("secondary tx not found, queue out of order")
```

Comment:
These panics seem out to place me, simply because the begin tx job could fail and the `SecondaryTx.Tx` would still equal `nil` but there would be not programmer error...